fil_fungible_token/src/token/mod.rs

```diff
@@ -294,6 +304,32 @@ where
         Ok(())
     }
 
+    /// Sets the allowance to a specified amount, returning the old allowance
+    pub fn set_allowance(
+        &mut self,
+        owner: &Address,
+        operator: &Address,
+        amount: &TokenAmount,
+    ) -> Result<TokenAmount> {
+        if amount.is_negative() {
+            return Err(TokenError::InvalidNegative { name: "allowance", amount: amount.clone() });
+        }
+
+        // Handle special revoke allowance case
+        if amount.is_zero() {
+            let old_allowance = self.allowance(owner, operator)?;
+            self.revoke_allowance(owner, operator)?;
+            return Ok(old_allowance);
+        }
```

Comment:
1) I think you need to resolve the addresses before looking up the allowance (can you write a test that fails here?)
2) The state method already does this, why is it repeated here?

Reply:
`revoke_allowance` and `allowance` have special shortcut cases which avoid initialising accounts that don't need to be initialised. So I delegate the logic of checking that to the existing methods. 

however, this means in cases where we know amount is being set to non-zero we have a duplicate sign check at the library and state level. I think this is ok as a defensive measure - since other sites can call set_allowance without checking sign.

Reply:
an alternative factoring could repeat the special handling of account-initialisation here but it would likely be a bit more verbose.

Reply:
or - perhaps for functionality beyond the standard interface, we should force consumers to operate directly on a `&mut` of State. this potentially serves also to highlight that extra care should be taken to maintain desired invariants.